clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,223 @@
+  // expected-note@#USN_F4{{'__builtin_unique_stable_name' evaluated here}}
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4); // #USN_F4
+
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel4func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel13_14func(const Func &F) {
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l11{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@#USN_l12{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel13>(F);
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  kernel_single_task<class kernel14>(F);
+  // Using the same functor twice should be fine
+}
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12); // #USN_l12
+// expected-note@-1{{expanded from macro 'MACRO12'}}
+
+int main() {
+
+  // kernel0
+  kernel_single_task<class kernel0>(func<Derp>);
+
+  // kernel1 - expect no error
+  // Evaluate the builtin in a constant context, then call the kernel with
+  // that same type
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); });
+
+  // kernel2 - expect error
+  // Call to the builtin on a lambda, then using the lambda for a kernel name
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2);
+
+  // kernel3 - expect error
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
```

Comment:
Again, the error-followed-by-notes should ALL be in 1 place.  Having these notes in the middle of nowhere is confusing.